### PR TITLE
Adding tests for RTCDataChannel id attribute.

### PR DESCRIPTION
**Pull request by taylor-b**
When no ID argument is provided, and the DTLS role hasn't been
determined, the id attribute should return `null`.

After the DTLS role is determined by an answer being applied with
"a=setup:active" (or "passive"), the null IDs should be replaced with
either odd or even IDs, depending on the negotiated role.

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/5742)

### CODE REVIEW

**Comment by taylor-b** [comment]
This is the thing @foolip requested on https://github.com/w3c/webrtc-pc/pull/1137, with an additional test for the odd/even ID selection after `setRemoteDescription`.

**Comment by agouaillard** [comment]
Read the thread, looked at the code. LGTM. As soon as folio approve, (and travis is happy) I will merge.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/227285403)

# Lint #

Passed

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/227285405)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 43bcbca9214c7ff34a2a5c3aab37fa41f4e00e93
Using browser at version BuildID 20170429100354; SourceStamp 308cdb913d717f8edc932f5bac93854e3e54b84d
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5742/webrtc/RTCPeerConnection-createDataChannel.html">/webrtc/RTCPeerConnection-createDataChannel.html</a></summary>


|                              Subtest                               | Results |                                                                                                                                                  Messages                                                                                                                                                  |
|--------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                    | OK      |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel required arguments`                             | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with closed connection`                         | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel defaults`                                       | FAIL    | `RTCDataChannel is not defined`                                                                                                                                                                                                                                                                            |
| `DTLS client uses odd data channel IDs`                            | FAIL    | `assert_unreached: Error Error: assert_equals: id expected 1 but got 65535 Reached unreachable code`                                                                                                                                                                                                       |
| `DTLS server uses even data channel IDs`                           | FAIL    | `assert_unreached: Error Error: assert_equals: id expected 0 but got 65535 Reached unreachable code`                                                                                                                                                                                                       |
| `createDataChannel with label "foo"`                               | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with label null`                                | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with label undefined`                           | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with label lone surrogate`                      | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with ordered false`                             | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with ordered null/undefined`                    | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with maxPacketLifeTime 0`                       | FAIL    | `assert_equals: expected (number) 0 but got (undefined) undefined`                                                                                                                                                                                                                                         |
| `createDataChannel with maxRetransmits 0`                          | FAIL    | `assert_equals: expected (number) 0 but got (undefined) undefined`                                                                                                                                                                                                                                         |
| `createDataChannel with both maxPacketLifeTime and maxRetransmits` | FAIL    | `assert_throws: function "() => pc.createDataChannel('', {\n    maxPacketLifeTime: 0,\n    maxRetransmits: 0\n  })" threw object "InvalidParameterError: Both maxPacketLifeTime and maxRetransmits cannot be provided" that is not a DOMException SyntaxError: property "code" is equal to 0, expected 12` |
| `createDataChannel with protocol "foo"`                            | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with protocol null`                             | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with protocol undefined`                        | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with protocol lone surrogate`                   | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with negotiated true`                           | FAIL    | `assert_equals: expected (boolean) true but got (undefined) undefined`                                                                                                                                                                                                                                     |
| `createDataChannel with id -1`                                     | FAIL    | `assert_throws: function "() => pc.createDataChannel('', { id: id })" did not throw`                                                                                                                                                                                                                       |
| `createDataChannel with id 0`                                      | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with id 1`                                      | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with id 65534`                                  | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with id 65535`                                  | FAIL    | `assert_throws: function "() => pc.createDataChannel('', { id: id })" did not throw`                                                                                                                                                                                                                       |
| `createDataChannel with id 65536`                                  | FAIL    | `assert_throws: function "() => pc.createDataChannel('', { id: id })" did not throw`                                                                                                                                                                                                                       |
| `createDataChannel with priority "high"`                           | FAIL    | `assert_equals: expected (string) "high" but got (undefined) undefined`                                                                                                                                                                                                                                    |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/227285411)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 43bcbca9214c7ff34a2a5c3aab37fa41f4e00e93

**Comment by taylor-b** [comment]
The new tests are failing on Firefox because it looks like Firefox only assigns the ID after the DTLS (and SCTP?) handshake completes. But the spec currently says it should be assigned by setting an answer. This was discussed briefly [in a virtual interim](https://www.w3.org/2017/03/01-webrtc-minutes#item03), and changed in this PR: https://github.com/w3c/webrtc-pc/pull/1038

But there wasn't a lot of discussion. @jesup, @jan-ivar, any opinions? Should the spec be more loose about when the ID is assigned? I

**Comment by agouaillard** [comment]
This is a spec question. If the test as is tests the current spec draft, I am tempted to merge it in, and let the discussion happen in the spec repository.

 If / when there is a spec change, then this test will be modified accordingly. 

The risk is that any test could be otherwise stay on hold waiting for the final version of the spec.

Unless there is a strong objection, I will merge it Tuesday evening end of day pacific time.

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (jan-ivar) [inline webrtc/RTCPeerConnection-createDataChannel.html]: @agouaillard By that standard we shouldn't use the legacy [RTCSessionDescription constructor](http://w3c.github.io/webrtc-pc/#dom-rtcsessiondescription) ;)

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (jesup) [inline webrtc/RTCPeerConnection-createDataChannel.html]: I don't think the spec says the next channel must be '2', just that it must be even.

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (jesup) [inline webrtc/RTCPeerConnection-createDataChannel.html]: I don't think the spec says this must be 0, just that it must be even

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (jesup) [inline webrtc/RTCPeerConnection-createDataChannel.html]: I don't think the spec says this must be 1, just that it must be odd:
"If the side is the DTLS client, it MUST choose an even
   Stream Identifier, if the side is the DTLS server, it MUST choose an
   odd one.  "

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (jesup) [inline webrtc/RTCPeerConnection-createDataChannel.html]: I don't think the spec says this must be 3, just that it must be odd

**Review by jesup — CHANGES_REQUESTED** [review]

**Comment by jesup** [comment]
The interim discussion was on "what do we return before the id is established", and agreed null made more sense than the current impl's 65535.

Assigning the ID when setting the answer would require we inform the DataChannel code of the DTLS roles that *will* be negotiated, assuming one is guaranteed to know that at Answer time.  Right now as stated we (firefox) assign them on DTLS negotiation, which gates us being able to actually send the Open packets on a specific id.

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5742/)

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (taylor-b) [inline webrtc/RTCPeerConnection-createDataChannel.html]: Fixed. Now only testing that the IDs are odd, and distinct from each other.

**Comment by taylor-b** [comment]
@jesup 

> Assigning the ID when setting the answer would require we inform the DataChannel code of the DTLS roles that will be negotiated, assuming one is guaranteed to know that at Answer time.

Since the answer contains the answerer's selected role, this should be guaranteed. Would it be feasible to change Firefox's behavior? Or should the spec's requirements be made more loose (for example, "the ID will be assigned at some point before `readyState` changes to `"open"`")?

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (foolip) [inline webrtc/RTCPeerConnection-createDataChannel.html]: This test would probably be slightly smaller using `promise_test`, which would avoid the need for this.

File: webrtc/RTCPeerConnection-createDataChannel.html
Comment (foolip) [inline webrtc/RTCPeerConnection-createDataChannel.html]: This could be `offer => pc.setLocalDescription(offer)`

**Review by foolip — CHANGES_REQUESTED** [review]

**Comment by taylor-b** [comment]
@foolip Addressed your comments; take another look.

**Review by foolip — APPROVED** [review]

**Pull request merged**
